lib/Dialect/FIRRTL/Transforms/Vectorization.cpp

```diff
@@ -69,7 +69,70 @@ class VectorCreateToLogicElementwise : public mlir::RewritePattern {
     return failure();
   }
 };
-} // namespace
+
+template <typename OpTy, typename ResultOpType>
+class ChainedReducer : public mlir::RewritePattern {
+public:
+  ChainedReducer(MLIRContext *context)
+      : RewritePattern(OpTy::getOperationName(), 0, context) {}
+
+  bool recurse(OpTy root, Value vec, DenseSet<size_t> &indexes) const {
+    auto lhsSub =
+        dyn_cast_or_null<SubindexOp>(root->getOperand(0).getDefiningOp());
+    auto rhsSub =
+        dyn_cast_or_null<SubindexOp>(root->getOperand(1).getDefiningOp());
+    auto lhsOp = dyn_cast_or_null<OpTy>(root->getOperand(0).getDefiningOp());
+    auto rhsOp = dyn_cast_or_null<OpTy>(root->getOperand(1).getDefiningOp());
+    // op(subindex(vec,x), op_chain);
+    // op(op_chain, subindex(vec,x))
+    // op(subindex(vec,x), subindex(vec,y))
+    if (lhsSub && rhsOp && lhsSub.getInput() == vec) {
+      indexes.insert(lhsSub.getIndex());
```

Comment:
For `xor` if one element is `xor`'d in more than once, will this do the right thing?
(for other operations this is okay I think)

A template parameter indicating whether the indices must be present exactly once (causing this to be checked) might be useful (perhaps "isIdempotent" if that seems right to you), or handle however.

Reply:
Good catch! Yes, this isn't quite right with `xor`.